backend/attachRendererFiber.js

```diff
@@ -0,0 +1,252 @@
+    if (!Array.isArray(prevChildren) || !Array.isArray(nextChildren)) {
+      return prevChildren !== nextChildren;
+    }
+    if (prevChildren.length !== nextChildren.length) {
+      return true;
+    }
+    for (let i = 0; i < prevChildren.length; i++) {
+      if (prevChildren[i] !== nextChildren[i]) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  function hasDataChanged(prevData, nextData) {
+    return (
+      prevData.ref !== nextData.ref ||
+      prevData.source !== nextData.source ||
+      prevData.props !== nextData.props ||
+      prevData.state !== nextData.state ||
+      prevData.context !== nextData.context ||
+      prevData.text !== nextData.text ||
+      haveChildrenChanged(prevData.children, nextData.children)
+    );
+  }
+
+  let pendingEvents = [];
+
+  function flushPendingEvents() {
+    const events = pendingEvents;
+    pendingEvents = [];
+    for (let i = 0; i < events.length; i++) {
+      const event = events[i];
+      hook.emit(event._event, event);
+    }
+  }
+
+  function enqueueMount(fiber) {
+    pendingEvents.push({
+      // TODO: the naming is confusing. `element` is *not* a React element. It is an opaque ID.
+      element: getOpaqueNode(fiber),
+      data: getDataFiber(fiber, getOpaqueNode),
+      renderer: rid,
+      _event: 'mount',
+    });
+
+    const isRoot = fiber.tag === 3;
+    if (isRoot) {
+      pendingEvents.push({
+        element: getOpaqueNode(fiber),
+        renderer: rid,
+        _event: 'root',
+      });
+    }
+  }
+
+  function enqueueUpdateIfNecessary(fiber) {
+    const nextData = getDataFiber(fiber, getOpaqueNode);
+    const prevData = getDataFiber(fiber.alternate, getOpaqueNode);
+    // Avoid unnecessary updates since they are common
+    // when something changes deep in the tree due to setState.
+    if (!hasDataChanged(prevData, nextData)) {
+      return;
+    }
+    pendingEvents.push({
+      element: getOpaqueNode(fiber),
+      data: nextData,
+      renderer: rid,
+      _event: 'update',
+    });
+  }
+
+  function enqueueUnmount(fiber) {
+    const isRoot = fiber.tag === 3;
+    const event = {
+      element: getOpaqueNode(fiber),
+      renderer: rid,
+      _event: 'unmount',
+    };
+    if (isRoot) {
+      pendingEvents.push(event);
+    } else {
+      // Non-root fibers are deleted during the commit phase.
+      // They are deleted in the child-first order. However
+      // DevTools currently expects deletions to be parent-first.
+      // This is why we unshift deletions rather than push them.
+      pendingEvents.unshift(event);
+    }
+    opaqueNodes.delete(fiber);
+    if (fiber.alternate != null) {
+      opaqueNodes.delete(fiber.alternate);
+    }
+  }
+
+  function mountFiber(fiber) {
+    // Depth-first.
+    // Logs mounting of children first, parents later.
+    let node = fiber;
+    outer: while (true) {
+      if (node.child) {
+        node.child.return = node;
```

Comment:
> They happen if return points to alternate of the parent you're traversing.

Wouldn't this only happen if there was a bug in fiber?

I expect these `return` assignments in devtools are basically neutral since the pointers should be correct to begin with, but it still seems weird to see something _outside of_ `React` setting values on fibers.